File: RadixCyclicNN/radixnet/beam.py
"""Beam search over the graph: the K best *and* the K worst continuations in one prediction.

A state is a partial path ``(cost, chars_emitted, node, entry)`` where
``entry`` indexes a parent-pointer table, so every partial path is a distinct
text.  The *top* beam keeps the ``beam`` cheapest partial paths per step, the
*bottom* beam the ``beam`` most expensive ones.  A path is complete at END,
once it emitted ``min_chars`` characters (unless ``to_end``), or at a cap.
``beam_predict`` returns the ``k`` cheapest complete paths, the ``k`` most
expensive complete paths that are not among them, and the number of
expansions.  Costs are the same as Dijkstra's (``-log softmax`` per edge plus
``step_penalty``), so ``exp(-cost)`` is a path's probability.
"""

from __future__ import annotations

import heapq
import math
from dataclasses import dataclass, field

from .encoding import WINDOW
from .graph import END, START, RadixCyclicGraph
from .search import (
    LEAST_PUNISHED,
    CostFn,
    PathResult,
    _build_result,
    _start_emission,
    least_punished,
    onward,
    parse_traversal,
)

__all__ = ["Prediction", "beam_predict", "default_beam", "path_probability"]

_OV = WINDOW - 1


@dataclass(slots=True)
class Prediction(PathResult):
    """A :class:`PathResult` (the best path) plus the top-K and bottom-K paths of the same search."""

    top: list[PathResult] = field(default_factory=list)
    bottom: list[PathResult] = field(default_factory=list)
    k: int = 0
    beam: int = 0
    mode: str = "beam"
    traversal: str = "reward"
    """Which traversal wrote this: ``"reward"``, ``"punishment"`` (:mod:`radixnet.penalty` prices the steps)
    or ``"least-punished"`` (``../SPEC-LeastPunished.md`` ranks the walks)."""

    def to_dict(self) -> dict:
        d = PathResult.to_dict(self)
        d.update(
            top=[p.to_dict() for p in self.top],
            bottom=[p.to_dict() for p in self.bottom],
            k=self.k,
            beam=self.beam,
            mode=self.mode,
            traversal=self.traversal,
        )
        if self.traversal and self.traversal != "reward":
            d["traversal"] = self.traversal
        return d


def default_beam(k: int) -> int:
    """Beam width used when none is given: wide enough for ``k`` distinct paths per side."""
    return max(4 * max(1, k), 16)


def path_probability(result: PathResult) -> float:
    """``exp(-cost)`` of a path (0 for an infinite cost)."""
    return math.exp(-result.cost) if math.isfinite(result.cost) else 0.0


def _run_beam(
    graph: RadixCyclicGraph,
    start_node: int,
    start_chars: int,
    min_chars: int,
    cap: int | None,
    k: int,
    width: int,
    step_penalty: float,
    to_end: bool,
    max_steps: int,
    max_expansions: int,
    worst: bool,
    traversal: str = "reward",
    costs: CostFn | None = None,
) -> tuple[list[tuple[float, list[int], list[float], float]], int]:
    """One beam: the ``k`` cheapest (or, with ``worst``, dearest) complete paths
    as ``(cost, node_ids, step_costs, punish)``.

    ``costs`` replaces the graph's own cost function, which is how the
    punishment traversal prices a step (:mod:`radixnet.penalty`).  Under
    ``traversal="least-punished"`` "cheapest" instead reads as "least punished,
    and cheapest among those": a path is ranked by its *worst* step first and by
    its summed cost only where two paths carry the same worst step.
    """
    labels = graph.labels
    blamed = traversal == LEAST_PUNISHED
    child_costs = graph.child_steps if blamed else (graph.child_costs if costs is None else costs)
    entries: list[tuple[int, int, float]] = [(start_node, -1, 0.0)]  # entry -> (node, parent, step cost)
    sign = -1.0 if worst else 1.0  # heap keys: the k-th best finished path sits at the heap top

    def complete(node: int, chars: int) -> bool:
        if node == END:
            return True
        if cap is not None and chars >= cap:
            return True
        return not to_end and chars >= min_chars

    # a bounded heap of (sign * -punish, sign * -cost, entry): the worst kept path
    # is at the top, and under the reward traversal the first component is 0.0 on
    # every path, so the whole comparison is the cost order the beam has always used
    done: list[tuple[float, float, int]] = []

    def offer(punish: float, cost: float, entry: int) -> None:
        if k == 0:
            return
        key = (-punish * sign if blamed else 0.0, -cost * sign, entry)
        if len(done) < k:
            heapq.heappush(done, key)
        elif key[:2] > done[0][:2]:
            heapq.heapreplace(done, key)

    start = (0.0, 0.0, start_chars, start_node, 0)
    frontier: list[tuple[float, float, int, int, int]] = []
    if complete(start_node, start_chars):
        offer(0.0, 0.0, 0)
    else:
        frontier.append(start)
    fallback = frontier
    expanded = 0
    steps = 0
    while frontier and steps < max_steps and expanded < max_expansions:
        steps += 1
        candidates: list[tuple[float, float, int, int, int]] = []
        for punish, cost, chars, node, entry in frontier:
            expanded += 1
            parent_entry = entries[entry][1]
            # where the walk came from: a model that counts paths prices the next step by it
            prev = entries[parent_entry][0] if parent_entry >= 0 else (START if node == START else None)
            children = onward(child_costs(node, prev))
            if blamed:
                children = least_punished(children)
            for item in children:
                c, ec = item[0], item[2]
                nchars = chars if c == END else chars + len(labels[c]) - _OV
                step = ec + step_penalty
                ncost = cost + step
                npunish = max(punish, item[3]) if blamed else 0.0  # a path is as punished as its worst step
                child_entry = len(entries)
                entries.append((c, entry, step))
                if complete(c, nchars):
                    offer(npunish, ncost, child_entry)
                else:
                    candidates.append((npunish, ncost, nchars, c, child_entry))
            if expanded >= max_expansions:
                break
        if candidates:
            fallback = candidates
        frontier = heapq.nlargest(width, candidates) if worst else heapq.nsmallest(width, candidates)
        # neither number shrinks along a path - a cost is a sum of costs and a punishment the
        # worst step so far - so once k paths finished and the best partial one already ranks
        # below the k-th of them, the best side is settled
        if not worst and k > 0 and len(done) == k and frontier:
            head = (-frontier[0][0] * sign if blamed else 0.0, -frontier[0][1] * sign)
            if head <= tuple(done[0][:2]):
                break

    def path_of(entry: int) -> tuple[list[int], list[float]]:
        node_ids: list[int] = []
        step_costs: list[float] = []
        while entry >= 0:
            node, parent, step = entries[entry]
            node_ids.append(node)
            if parent >= 0:
                step_costs.append(step)
            entry = parent
        node_ids.reverse()
        step_costs.reverse()
        return node_ids, step_costs

    if done:
        # 0.0 - x rather than -x: a negative zero reads as a punishment that is not there
        kept = [(0.0 - punish * sign, 0.0 - cost * sign, entry) for punish, cost, entry in done]
        finished = sorted(kept, key=lambda item: (item[0] * sign if blamed else 0.0, item[1] * sign, item[2]))
    elif fallback:
        # nothing completed within the limits: the surviving partial paths, most characters first
        pick = heapq.nsmallest(
            k, fallback, key=lambda s: (-s[2], s[0] * sign if blamed else 0.0, s[1] * sign, s[4])
        )
        finished = [(s[0], s[1], s[4]) for s in pick]
    else:
        finished = []
    return [(cost, *path_of(entry), punish) for punish, cost, entry in finished], expanded


def beam_predict(
    graph: RadixCyclicGraph,
    start_node: int,
    start_offset: int,
    min_chars: int,
    k: int = 5,
    beam: int | None = None,
    max_chars: int | None = None,
    step_penalty: float = 0.0,
    to_end: bool = False,
    max_steps: int | None = None,
    max_expansions: int = 200_000,
    include_context: bool | None = None,
    traversal: str = "reward",
    costs: CostFn | None = None,
) -> tuple[list[PathResult], list[PathResult], int]:
    """``(top, bottom, expanded)``: the ``k`` cheapest and the ``k`` most expensive complete paths.

    ``top`` is sorted by rising cost, ``bottom`` by falling cost, and no path
    appears in both.  With ``to_end`` and no ``max_chars`` the bottom side is
    capped at twice the length of the longest top path (plus a margin), so the
    "worst" paths stay comparable instead of cycling for hundreds of steps.
    When nothing completes within ``max_steps`` rounds (default ``min_chars +
    50``, or 500 with ``to_end``) or ``max_expansions``, the surviving partial
    paths are returned instead (``reached_end`` is ``False`` on them).  A start
    that already satisfies the goal gives one complete path (the start itself).
    ``costs`` replaces the graph's own cost function: with the punishment
    traversal's (:mod:`radixnet.penalty`) ``top`` is the ``k`` *least punished*
    continuations and ``bottom`` the ``k`` most punished ones.
    ``traversal="least-punished"`` does something else again: it ranks a path by
    the blame on its worst step before its cost, and lets a node offer only the
    children it has the least against (``../SPEC-LeastPunished.md``).
    """
    traversal = parse_traversal(traversal)
    if k < 0:
        raise ValueError(f"k must be >= 0, got {k}")
    if step_penalty < 0:
        raise ValueError("step_penalty must be >= 0")
    width = default_beam(k) if beam is None else int(beam)
    if width < 1:
        raise ValueError(f"beam must be >= 1, got {beam}")
    if max_chars is not None and max_chars < min_chars:
        max_chars = min_chars
    if max_steps is None:
        max_steps = 500 if to_end else min_chars + 50
    start_chars = _start_emission(graph, start_node, start_offset)
    if k == 0:
        return [], [], 0
    common = (graph, start_node, start_chars, min_chars)
    limits = (k, width, step_penalty, to_end, max_steps, max_expansions)
    best, expanded = _run_beam(*common, max_chars, *limits, False, traversal, costs)
    bottom_cap = max_chars
    if bottom_cap is None and to_end:
        longest = max((len(graph.labels[n]) for _, ids, _, _ in best for n in ids[1:]), default=0)
        emitted = max(
            (sum(len(graph.labels[n]) - _OV for n in ids[1:] if n != END) for _, ids, _, _ in best), default=0
        )
        bottom_cap = max(2 * emitted + longest + 8, min_chars, 16)
    worst, expanded_worst = _run_beam(*common, bottom_cap, *limits, True, traversal, costs)
    expanded += expanded_worst
    seen = {tuple(ids) for _, ids, _, _ in best}

    def built(rows: list, cap: int | None) -> list[PathResult]:
        out = []
        for _cost, ids, steps, punish in rows:
            result = _build_result(graph, ids, steps, start_offset, cap, expanded, include_context)
            result.punish = punish
            out.append(result)
        return out

    top = built(best, max_chars)
    bottom = built([row for row in worst if tuple(row[1]) not in seen], bottom_cap)
    return top, bottom, expanded

File: RadixCyclicNN/go/radixnet/beam.go
package radixnet

import (
	"fmt"
	"sort"
	"sync"
)

// Prediction is the best path plus the top-K and bottom-K paths of one search.
type Prediction struct {
	PathResult
	Top    []*PathResult `json:"top"`
	Bottom []*PathResult `json:"bottom"`
	K      int           `json:"k"`
	Beam   int           `json:"beam"`
	Mode   string        `json:"mode"`
	// Traversal names the search that produced this, when it was not the
	// ordinary one by cost.
	Traversal string `json:"traversal,omitempty"`
}

// DefaultBeam is the beam width used when none is given.
func DefaultBeam(k int) int {
	if k < 1 {
		k = 1
	}
	if 4*k > 16 {
		return 4 * k
	}
	return 16
}

type beamState struct {
	cost   float64
	chars  int
	node   int
	entry  int
	punish float64 // the worst step on the way here (ByLeastPunished)
}

type beamEntry struct {
	node, parent int
	step         float64
}

type finished struct {
	cost   float64
	ids    []int
	step   []float64
	punish float64
}

func lessState(a, b beamState, traversal Traversal) bool {
	if traversal == ByLeastPunished && a.punish != b.punish {
		return a.punish < b.punish // the walk that has the least against it leads
	}
	if a.cost != b.cost {
		return a.cost < b.cost
	}
	if a.chars != b.chars {
		return a.chars < b.chars
	}
	if a.node != b.node {
		return a.node < b.node
	}
	return a.entry < b.entry
}

// runBeam is one beam: the k cheapest (or, with worst, dearest) complete paths.
// Under ByLeastPunished "cheapest" reads as "least punished, and cheapest among
// those": a path is ranked by its worst step first and by its summed cost only
// where two paths carry the same worst step.
func runBeam(g *Graph, startNode, startChars, minChars, cap, k, width int, stepPenalty float64, toEnd bool, maxSteps, maxExpansions int, worst bool, traversal Traversal) ([]finished, int) {
	entries := []beamEntry{{startNode, -1, 0}}
	sign := 1.0
	if worst {
		sign = -1.0
	}
	complete := func(node, chars int) bool {
		if node == End {
			return true
		}
		if cap >= 0 && chars >= cap {
			return true
		}
		return !toEnd && chars >= minChars
	}
	// A kept path is two numbers, higher is better in this beam's own direction:
	// what the walk was punished for and what it cost.  Under ByReward the first
	// is 0 on every path, so the whole comparison is the cost order the beam has
	// always used.
	type doneKey struct {
		first, second float64
		entry         int
	}
	keyOf := func(cost, punish float64, entry int) doneKey {
		first := 0.0
		if traversal == ByLeastPunished {
			first = -punish * sign
		}
		return doneKey{first, -cost * sign, entry}
	}
	betterThan := func(a, b doneKey) bool {
		if a.first != b.first {
			return a.first > b.first
		}
		return a.second > b.second
	}
	done := make([]doneKey, 0, k)
	worstIndex := func() int { // the kept path that ranks last: the first to be replaced
		best := 0
		for i := 1; i < len(done); i++ {
			a, b := done[i], done[best]
			if a.first != b.first {
				if a.first < b.first {
					best = i
				}
				continue
			}
			if a.second != b.second {
				if a.second < b.second {
					best = i
				}
				continue
			}
			if a.entry < b.entry {
				best = i
			}
		}
		return best
	}
	offer := func(cost, punish float64, entry int) {
		if k == 0 {
			return
		}
		key := keyOf(cost, punish, entry)
		if len(done) < k {
			done = append(done, key)
			return
		}
		w := worstIndex()
		if betterThan(key, done[w]) {
			done[w] = key
		}
	}
	frontier := []beamState{}
	start := beamState{0, startChars, startNode, 0, 0}
	if complete(startNode, startChars) {
		offer(0, 0, 0)
	} else {
		frontier = append(frontier, start)
	}
	fallback := frontier
	expanded, steps := 0, 0
	for len(frontier) > 0 && steps < maxSteps && expanded < maxExpansions {
		steps++
		candidates := make([]beamState, 0, len(frontier)*2)
		for _, st := range frontier {
			expanded++
			// where the walk came from: a judged path prices the next step by it
			prev := -1
			if parent := entries[st.entry].parent; parent >= 0 {
				prev = entries[parent].node
			} else if st.node == Start {
				prev = Start
			}
			children := Onward(g.ChildCostsFrom(st.node, prev))
			if traversal == ByLeastPunished {
				children = LeastPunished(children)
			}
			for _, cc := range children {
				nchars := st.chars
				if cc.Child != End {
					nchars += g.labelLen[cc.Child] - Overlap
				}
				step := cc.Cost + stepPenalty
				ncost := st.cost + step
				npunish := st.punish
				if cc.Punish > npunish {
					npunish = cc.Punish // a path is as punished as its worst step
				}
				childEntry := len(entries)
				entries = append(entries, beamEntry{cc.Child, st.entry, step})
				if complete(cc.Child, nchars) {
					offer(ncost, npunish, childEntry)
				} else {
					candidates = append(candidates, beamState{ncost, nchars, cc.Child, childEntry, npunish})
				}
			}
			if expanded >= maxExpansions {
				break
			}
		}
		if len(candidates) > 0 {
			fallback = candidates
		}
		if worst {
			sort.Slice(candidates, func(i, j int) bool { return lessState(candidates[j], candidates[i], traversal) })
		} else {
			sort.Slice(candidates, func(i, j int) bool { return lessState(candidates[i], candidates[j], traversal) })
		}
		if len(candidates) > width {
			candidates = candidates[:width]
		}
		frontier = candidates
		// neither number shrinks along a path - a cost is a sum of costs and a
		// punishment the worst step so far - so once k paths finished and the best
		// partial one already ranks below the k-th of them, the best side is settled
		if !worst && k > 0 && len(done) == k && len(frontier) > 0 {
			w := worstIndex()
			head := keyOf(frontier[0].cost, frontier[0].punish, frontier[0].entry)
			if !betterThan(head, done[w]) {
				break
			}
		}
	}
	pathOf := func(entry int) ([]int, []float64) {
		ids := []int{}
		stepsOut := []float64{}
		for entry >= 0 {
			en := entries[entry]
			ids = append(ids, en.node)
			if en.parent >= 0 {
				stepsOut = append(stepsOut, en.step)
			}
			entry = en.parent
		}
		for i, j := 0, len(ids)-1; i < j; i, j = i+1, j-1 {
			ids[i], ids[j] = ids[j], ids[i]
		}
		for i, j := 0, len(stepsOut)-1; i < j; i, j = i+1, j-1 {
			stepsOut[i], stepsOut[j] = stepsOut[j], stepsOut[i]
		}
		return ids, stepsOut
	}
	type pick struct {
		cost   float64
		punish float64
		entry  int
	}
	var picks []pick
	if len(done) > 0 {
		for _, d := range done {
			punish := -d.first * sign
			if punish == 0 {
				punish = 0 // a negative zero reads as a punishment that is not there
			}
			picks = append(picks, pick{-d.second * sign, punish, d.entry})
		}
		sort.Slice(picks, func(i, j int) bool {
			if traversal == ByLeastPunished && picks[i].punish != picks[j].punish {
				return picks[i].punish*sign < picks[j].punish*sign
			}
			a, b := picks[i].cost*sign, picks[j].cost*sign
			if a != b {
				return a < b
			}
			return picks[i].entry < picks[j].entry
		})
	} else if len(fallback) > 0 {
		// nothing completed within the limits: the surviving partial paths, most characters first
		fb := append([]beamState(nil), fallback...)
		sort.Slice(fb, func(i, j int) bool {
			if fb[i].chars != fb[j].chars {
				return fb[i].chars > fb[j].chars
			}
			if traversal == ByLeastPunished && fb[i].punish != fb[j].punish {
				return fb[i].punish*sign < fb[j].punish*sign
			}
			a, b := fb[i].cost*sign, fb[j].cost*sign
			if a != b {
				return a < b
			}
			return fb[i].entry < fb[j].entry
		})
		if len(fb) > k {
			fb = fb[:k]
		}
		for _, s := range fb {
			picks = append(picks, pick{s.cost, s.punish, s.entry})
		}
	}
	out := make([]finished, 0, len(picks))
	for _, p := range picks {
		ids, st := pathOf(p.entry)
		out = append(out, finished{p.cost, ids, st, p.punish})
	}
	return out, expanded
}

// BeamOptions configure BeamPredict.
type BeamOptions struct {
	K             int
	Beam          int // 0 = DefaultBeam(K)
	MaxChars      int // -1 = no cap
	StepPenalty   float64
	ToEnd         bool
	MaxSteps      int // 0 = default (500 with ToEnd, else MinChars + 50)
	MaxExpansions int // 0 = 200000
	// Traversal is how the two beams rank a path: ByReward (the default) by
	// cost, ByLeastPunished by the blame on its worst step first.
	Traversal Traversal
}

// BeamPredict returns the k cheapest complete paths (rising cost) and the k
// most expensive ones not among them (falling cost), plus the expansions.
// The two beams run concurrently when the bottom side's cap is known up front.
func (g *Graph) BeamPredict(startNode, startOffset, minChars int, opts BeamOptions) ([]*PathResult, []*PathResult, int, error) {
	if opts.K < 0 {
		return nil, nil, 0, fmt.Errorf("k must be >= 0, got %d", opts.K)
	}
	if opts.StepPenalty < 0 {
		return nil, nil, 0, fmt.Errorf("step_penalty must be >= 0")
	}
	width := opts.Beam
	if width == 0 {
		width = DefaultBeam(opts.K)
	}
	if width < 1 {
		return nil, nil, 0, fmt.Errorf("beam must be >= 1, got %d", opts.Beam)
	}
	maxChars := opts.MaxChars
	if maxChars >= 0 && maxChars < minChars {
		maxChars = minChars
	}
	maxSteps := opts.MaxSteps
	if maxSteps == 0 {
		if opts.ToEnd {
			maxSteps = 500
		} else {
			maxSteps = minChars + 50
		}
	}
	maxExp := opts.MaxExpansions
	if maxExp == 0 {
		maxExp = 200_000
	}
	g.Prepare()
	startChars, err := startEmission(g, startNode, startOffset)
	if err != nil {
		return nil, nil, 0, err
	}
	if opts.K == 0 {
		return []*PathResult{}, []*PathResult{}, 0, nil
	}
	var best, worstPaths []finished
	var expanded, expandedWorst int
	bottomCap := maxChars
	if (bottomCap < 0 && opts.ToEnd) || g.Workers == 1 {
		// the bottom cap depends on the best side, or one worker was asked for:
		// run the two beams in turn
		best, expanded = runBeam(g, startNode, startChars, minChars, maxChars, opts.K, width, opts.StepPenalty, opts.ToEnd, maxSteps, maxExp, false, opts.Traversal)
		if bottomCap < 0 && opts.ToEnd {
			longest, emitted := 0, 0
			for _, f := range best {
				sum := 0
				for _, n := range f.ids[1:] {
					if l := g.labelLen[n]; l > longest {
						longest = l
					}
					if n != End {
						sum += g.labelLen[n] - Overlap
					}
				}
				if sum > emitted {
					emitted = sum
				}
			}
			bottomCap = 2*emitted + longest + 8
			if minChars > bottomCap {
				bottomCap = minChars
			}
			if bottomCap < 16 {
				bottomCap = 16
			}
		}
		worstPaths, expandedWorst = runBeam(g, startNode, startChars, minChars, bottomCap, opts.K, width, opts.StepPenalty, opts.ToEnd, maxSteps, maxExp, true, opts.Traversal)
	} else {
		var wg sync.WaitGroup
		wg.Add(2)
		go func() {
			defer wg.Done()
			best, expanded = runBeam(g, startNode, startChars, minChars, maxChars, opts.K, width, opts.StepPenalty, opts.ToEnd, maxSteps, maxExp, false, opts.Traversal)
		}()
		go func() {
			defer wg.Done()
			worstPaths, expandedWorst = runBeam(g, startNode, startChars, minChars, bottomCap, opts.K, width, opts.StepPenalty, opts.ToEnd, maxSteps, maxExp, true, opts.Traversal)
		}()
		wg.Wait()
	}
	expanded += expandedWorst
	seen := make(map[string]bool, len(best))
	top := make([]*PathResult, 0, len(best))
	for _, f := range best {
		seen[fmt.Sprint(f.ids)] = true
		result := buildResult(g, f.ids, f.step, startOffset, maxChars, expanded, nil)
		result.Punish = f.punish
		top = append(top, result)
	}
	bottom := make([]*PathResult, 0, len(worstPaths))
	for _, f := range worstPaths {
		if !seen[fmt.Sprint(f.ids)] {
			result := buildResult(g, f.ids, f.step, startOffset, bottomCap, expanded, nil)
			result.Punish = f.punish
			bottom = append(bottom, result)
		}
	}
	return top, bottom, expanded, nil
}
